apps/aehttp/src/sc_ws_handler.erl

```diff
@@ -52,46 +85,125 @@ websocket_init(Params) ->
         {Handler, ChannelOpts} ->
             lager:debug("Starting Channel WS with params ~p",
                 [aesc_utils:censor_init_opts(Params)]),
-            lager:debug("ChannelOpts = ~p", [aesc_utils:censor_init_opts(ChannelOpts)]),
-            case start_link_fsm(Handler, ChannelOpts) of
-                {ok, FsmPid} ->
-                    MRef = erlang:monitor(process, FsmPid),
-                    {ok, Handler#handler{fsm_pid = FsmPid, fsm_mref = MRef}};
-                {error, Err} ->
-                    handler_init_error(Err, Handler)
-            end
+            websocket_init_parsed(Handler, ChannelOpts)
+    end.
+
+websocket_init_parsed(Handler, ChannelOpts) ->
+    lager:debug("ChannelOpts = ~p", [aesc_utils:censor_init_opts(ChannelOpts)]),
+    case start_link_fsm(Handler, ChannelOpts) of
+        {ok, FsmPid} ->
+            MRef = erlang:monitor(process, FsmPid),
+            {ok, Handler#handler{fsm_pid = FsmPid, fsm_mref = MRef}};
+        {error, Err} ->
+            handler_init_error(Err, Handler)
     end.
 
-websocket_init_reconnect(#{ <<"reconnect_tx">> := ReconnectTx } = Params) ->
+websocket_init_reconnect(#{ job_id := JobId
+                          , <<"protocol">> := Protocol
+                          , <<"reconnect_tx">> := ReconnectTx } = Params0) ->
+    Read = sc_ws_utils:read_f(Params0),
+    Params = sc_ws_utils:check_params(read_state_password(Read)),
     lager:debug("ReconnectTx = ~p", [ReconnectTx]),
     case check_reconnect_tx(ReconnectTx) of
-        {ok, #{ channel_id := ChanId
-              , role       := Role
-              , pub_key    := Pubkey
-              , signed_tx  := SignedTx } = Opts} ->
-            Handler = reconnect_opts_to_handler(maps:merge(Params, Opts)),
-            case aesc_fsm:where(ChanId, Role) of
-                undefined ->
-                    lager:debug("where(~p, ~p) -> undefined", [ChanId, Role]),
-                    handler_init_error(not_found, Handler);
-                #{ fsm_pid := Fsm, pub_key := Pubkey } ->
-                    %% At this point, we haven't verified the signature.
-                    %% This is done by the fsm.
-                    lager:debug("Found FSM = ~p", [Fsm]),
-                    case aesc_fsm:reconnect_client(Fsm, self(), SignedTx) of
-                        ok ->
-                            MRef = erlang:monitor(process, Fsm),
-                            { ok, Handler#handler{ fsm_pid  = Fsm
-                                                 , fsm_mref = MRef } };
-                        {error, Err} ->
-                            handler_init_error(Err, Handler)
-                    end
+        {ok, #{} = Opts} ->
+            lager:debug("Reconnect tx Opts = ~p", [Opts]),
```

Comment:
This will leak the current user provided state password :) The protection against leakage kicks in only in the FSM process.